### ISSUE TITLE
Use case not covered by react 16.4 docs? Synchronous rendering and using a model driven by props to control state

### ISSUE THREAD

**Issue by johnjesse** [reactjs/reactjs.org#978]
I wasn't sure whether to post this here or on facebook/react. This feels like a fairly standard use case as people use controlled inputs all the time, but I haven't managed to come across anything online.

I have a component that manages some form state for a form control element, e.g. a text input. Instead of using the component state for state management it uses a separate model e.g. Mobx. Note that I'm not just copying props across to some state, I'm using it to derive some other state, about whether the value is valid, invalid - potentially pending on asynchronous validation etc...

Pre react 16.4 I would use the now deprecated componentWillReceiveProps to drive this model from a value prop. And moving to the next lifecycle methods would seem to suggest I have to use componentDidUpdate in order to update the model from the prop. But - if I remember correctly - the render is called synchronously after the event handler for the input change event - specifically to avoid flickers of input value - and therefore I can't wait for componentDidUpdate to update the model because that would result in the incorrect value being put in the input in between updates

In react 16.4 I can't see how I'd achieve this. If I'm using a separate model for the state of a component and driving that from props then I can only do that in componentDidUpdate, is that correct? If so how do I use this pattern to manage the state of components who depend on rendering synchronously?

**Comment by johnjesse** [comment]
I've added a brief example here https://codesandbox.io/s/7k2kvr3ylj, if you play with the two input boxes you'll notice the second one doesn't work properly if you do an edit in the middle of the value.

The examples a bit contrived but I think it conveys what I'm trying to do - namely use something else e.g. mobx for managing state, and updating that from component props.

**Comment by johnjesse** [comment]
I've had a bit of a look around and https://github.com/mobxjs/mobx-react/issues/425 and https://github.com/facebook/react/issues/12310 seem related

I'm happy with the general answer being  don't use componentWillReceiveProps use componentDidUpdate - it's just in the case where two renders cause an issue e.g. with a controlled text input.

So it looks to me like the answer, in this case, is either - 
1. Use setState and local component state - not a different model like mobx
2. Don't update the model from props, instead receive the model as a prop.

Is that correct?

**Comment by gaearon** [comment]
I think this is correct, yes.

In React "receiving" new props doesn't mean it's a good time to mutate a model. Specifically with upcoming [async rendering features](https://reactjs.org/blog/2018/03/01/sneak-peek-beyond-react-16.html), this approach wouldn't make sense: a component could "receive" some props, but then get interrupted by a higher priority update, or "suspended" because we're still loading data.

As long as you stay within React's paradigm this doesn't affect you, but mutating external models (like MobX) exposes those issues.

Try treating props more like function arguments: just because you received them, doesn't mean it's time to mutate something "for good".

This lifecycle diagram might help visualize it:
http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/

Don't miss the legend on the left that shows the separation between "render" phase (can be paused/restarted) and the "commit" phase (which guarantees changes have been applied).

**Comment by plonko** [comment]
@gaearon Hey that URL is down and the link in the docs is broken. 
https://reactjs.org/docs/react-component.html#the-component-lifecycle

Can start a new issue if needed
